sw/device/lib/dif/dif_hmac.c

```diff
@@ -0,0 +1,400 @@
+                                    offset);
+  } else {
+    mmio_region_nonatomic_clear_bit32(hmac->base_addr,
+                                      HMAC_INTR_STATE_REG_OFFSET, offset);
+  }
+
+  return kDifHmacOk;
+}
+
+dif_hmac_result_t dif_hmac_irq_force(const dif_hmac_t *hmac,
+                                     dif_hmac_interrupt_t irq_type) {
+  if (hmac == NULL) {
+    return kDifHmacBadArg;
+  }
+
+  ptrdiff_t offset;
+  if (!irq_bit_offset_get(irq_type, &offset)) {
+    return kDifHmacError;
+  }
+
+  // Force the requested interrupt.
+  mmio_region_nonatomic_set_bit32(hmac->base_addr, HMAC_INTR_TEST_REG_OFFSET,
+                                  offset);
+
+  return kDifHmacOk;
+}
+
+dif_hmac_result_t dif_hmac_mode_hmac_start(const dif_hmac_t *hmac,
+                                           const uint8_t *key) {
+  if (hmac == NULL || key == NULL) {
+    return kDifHmacBadArg;
+  }
+
+  // Disable SHA256 while configuring and to clear the digest.
+  mmio_region_nonatomic_clear_bit32(hmac->base_addr, HMAC_CFG_REG_OFFSET,
+                                    HMAC_CFG_SHA_EN);
+
+  load_key(hmac, key);
+
+  // Set HMAC to process in HMAC mode (not SHA256-only mode).
+  mmio_region_nonatomic_set_mask32(
+      hmac->base_addr, HMAC_CFG_REG_OFFSET,
+      (1 << HMAC_CFG_SHA_EN) | (1 << HMAC_CFG_HMAC_EN), 0);
+
+  // Begin HMAC operation.
+  mmio_region_nonatomic_set_bit32(hmac->base_addr, HMAC_CMD_REG_OFFSET,
+                                  HMAC_CMD_HASH_START);
+  return kDifHmacOk;
+}
+
+dif_hmac_result_t dif_hmac_mode_sha256_start(const dif_hmac_t *hmac) {
+  if (hmac == NULL) {
+    return kDifHmacBadArg;
+  }
+
+  // Disable SHA256 while configuring and to clear the digest.
+  mmio_region_nonatomic_clear_bit32(hmac->base_addr, HMAC_CFG_REG_OFFSET,
+                                    HMAC_CFG_SHA_EN);
+
+  // Disable HMAC mode.
+  mmio_region_nonatomic_clear_bit32(hmac->base_addr, HMAC_CFG_REG_OFFSET,
+                                    HMAC_CFG_HMAC_EN);
+
+  // Set HMAC to process in SHA256-only mode.
+  mmio_region_nonatomic_set_bit32(hmac->base_addr, HMAC_CFG_REG_OFFSET,
+                                  HMAC_CFG_SHA_EN);
+
+  // Begin SHA256-only operation.
+  mmio_region_nonatomic_set_bit32(hmac->base_addr, HMAC_CMD_REG_OFFSET,
+                                  HMAC_CMD_HASH_START);
+
+  return kDifHmacOk;
+}
+
+dif_hmac_fifo_result_t dif_hmac_fifo_push(const dif_hmac_t *hmac,
+                                          const void *data, size_t len,
+                                          size_t *bytes_sent) {
+  if (hmac == NULL || data == NULL) {
+    return kDifHmacFifoBadArg;
+  }
+
+  const uint8_t *data_sent = (const uint8_t *)data;
+  int bytes_remaining = len;
+
+  while (bytes_remaining > 0 && get_fifo_available_space(hmac) > 0) {
+    bool word_aligned = (uintptr_t)data_sent % sizeof(uint32_t) == 0;
+    int bytes_written = 0;
```

Comment:
I think, since `len` and `bytes_sent` are `size_t`, `bytes_remaining` and `bytes_written` should also be `size_t`.